### PR TITLE
PS-8810: FEDERATED Update on Table can silently fail when Table contains DOUBLE Columns

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PS-8810

Problem:
If the table contains DOUBLE column with specified precision, update of the row, even if DOUBLE column is not specified in WHERE clause, reports that row has been updated, but the row is not updated.

Cause:
Federated SE constructs the query in a way that specifies all columns in WHERE clause even if the original query didn't contain all queries.

If the row contains DOUBLE(M,D) column, the value in WHERE clause is rounded to D digits after the decimal point compared to the underlying, actually stored value. This is because of the field definition. This rounded value goes to the server side and appears there as Item_decimal. It is converted to double in the following way:
  1. convert decimal to a string by decimal2string()
  2. convert string to double by my_strtod() Then analyze_field_constant() detects that the result would be rounded for comparison with the field having precision (M,D), so it is decided that row matching is not safe and query execution does not continue.

Solution:
Force server side to pass stored value through string to double conversion which results in the same number as the conversion of provided constant to double (both go through the same conversion path).

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-pipeline-parallel-mtr/126/

File: storage/federated/ha_federated.cc [hunk 1]
```diff
@@ -2054,31 +2107,38 @@ int ha_federated::update_row(const uchar *old_data, uchar *) {
...
     if (bitmap_is_set(table->read_set, (*field)->field_index())) {
-      size_t field_name_length = strlen((*field)->field_name);
-      append_ident(&where_string, (*field)->field_name, field_name_length,
-                   ident_quote_char);
-      if ((*field)->is_null_in_record(old_data))
-        where_string.append(STRING_WITH_LEN(" IS NULL "));
-      else {
-        bool needs_quote = (*field)->str_needs_quotes();
-        where_string.append(STRING_WITH_LEN(" = "));
+      if ((*field)->type() == MYSQL_TYPE_DOUBLE &&
```
Comment (percona-ysorokin) [inline storage/federated/ha_federated.cc]: Do we need to do the same for `MYSQL_TYPE_FLOAT`?
Also `MYSQL_TYPE_DECIMAL` and `MYSQL_TYPE_NEWDECIMAL`

File: storage/federated/ha_federated.cc [hunk 2]
```diff
@@ -1957,6 +1957,59 @@ int ha_federated::repair(THD *, HA_CHECK_OPT *check_opt) {
...
+
+  This is desired behavior because if we have table:
+  CREATE TABLE t1 (id INT PRIMARY KEY AUTO_INCREMENT, i INT, d DOUBLE(8,6));
+  INSERT INTO f1(i, d) VALUES (0, 8.730550);
+  and we execute:
+  UPDATE t1 set i=1 where d=8.7305501;
+  we do not expect row to be updated.
+
+  Note that there is the same issue even if we use a single server.
+  If we DOUBLE(M,D) column is specified in WHERE clause, the column is row is
```
Comment (percona-ysorokin) [inline storage/federated/ha_federated.cc]:
```suggestion
  If DOUBLE(M,D) column is specified in WHERE clause, the column is
```

File: storage/federated/ha_federated.cc (hunk 1 above)
Comment (kamil-holubicki) [reply storage/federated/ha_federated.cc]: There is special handling of float type in analyze_real_field_constant(). It explicitly downcasts double value to float and compares floats. We may have a precision mismatch for double comparison, but as we downcast to float, values are equal.

For DECIMAL type if the constant provided in WHERE clause fits the field precision, it is represented internally as Item_decimal 'as is' without any conversions and roundups.

File: storage/federated/ha_federated.cc (hunk 2 above)
Comment (kamil-holubicki) [reply storage/federated/ha_federated.cc]: fixed

**Comment by kamil-holubicki** [comment]
As we discussed on Slack:
> may be we should use your new handle_double_type() ONLY if the field has this (M, D) specification
I've limited the fix only to the case where precision is specified explicitly. I've also extended the MTR test.

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**